Gems/Terrain/Assets/Shaders/Terrain/TerrainPBR_ForwardPass.azsl

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Atom/Features/SrgSemantics.azsli>
+#include <viewsrg.srgi>
+#include <TerrainCommon.azsli>
+#include <Atom/RPI/ShaderResourceGroups/DefaultDrawSrg.azsli>
+#include <Atom/Features/PBR/ForwardPassSrg.azsli>
+#include <Atom/Features/PBR/ForwardPassOutput.azsli>
+#include <Atom/Features/PBR/AlphaUtils.azsli>
+#include <Atom/Features/ColorManagement/TransformColor.azsli>
+#include <Atom/Features/PBR/Lighting/StandardLighting.azsli>
+#include <Atom/Features/Shadow/DirectionalLightShadow.azsli>
+#include <Atom/Features/PBR/Decals.azsli>
+
+struct VSOutput
+{
+    float4 m_position : SV_Position;
+    float3 m_normal: NORMAL;
+    float3 m_tangent : TANGENT; 
+    float3 m_bitangent : BITANGENT; 
+    float3 m_worldPosition : UV0;
+    float3 m_shadowCoords[ViewSrg::MaxCascadeCount] : UV2;
+    float2 m_uv : UV1;
+};
+
+VSOutput TerrainPBR_MainPassVS(VertexInput IN)
+{
+    VSOutput OUT;
+ 
+    ObjectSrg::TerrainData terrainData = ObjectSrg::m_terrainData;
+
+    float2 uv = IN.m_uv;
+    float2 origUv = lerp(terrainData.m_uvMin, terrainData.m_uvMax, uv);
+    float3 worldPosition = GetTerrainWorldPosition(terrainData, IN.m_position, origUv);
+    OUT.m_position = mul(ViewSrg::m_viewProjectionMatrix, float4(worldPosition, 1.0));
+    OUT.m_worldPosition = worldPosition;
+
+    // Calculate normal
+    float up    = GetHeight(origUv + terrainData.m_uvStep * float2( 0.0f, -1.0f));
+    float right = GetHeight(origUv + terrainData.m_uvStep * float2( 1.0f,  0.0f));
+    float down  = GetHeight(origUv + terrainData.m_uvStep * float2( 0.0f,  1.0f));
+    float left  = GetHeight(origUv + terrainData.m_uvStep * float2(-1.0f,  0.0f));
```

Comment:
I'm curious how expensive sampling the texture this many times is (especially with smoothing enabled). The world position/neighboring heights are being re-calculated, but not actually changing from frame to frame, correct?

This suggestion might not be worthwhile, depending on how much all of this is going to change, but if you're going to take this approach for a while, it might be worth kicking off a compute shader to calculate the height and store it in the model. It would only need to run once whenever a new lod is loaded, and it would only need to run once per-vertex. You could store the TBN if it's not prohibitive to pay the memory cost, or you could re-calculate it here, but do it by getting the height from the neighboring vertices (though then you'd have to manually deal with the edges of the sectors, which might be feasible if you stored only the normal, then computed the slope from the normal, and used that to get the z-component of the tangent/bitangent, eliminating the need to get neighboring vertices).

Reply:
If we keep this long term we'll probably want a way to cache the result and not calculate every frame. While not currently written this way, it really doesn't need to be a lot of work because it only needs to be done on the few 10k's of verts nearest the camera. 

Anyway, this was mostly to just see what kind of effect it had on the output to see if it was something worth considering. I just stole the function from some work I'd done for LUTs and plugged it in (with modifications for a single channel Texture2D instead of a 3 channel Texture3D). It looks really nice, but also wouldn't line up perfectly with physics. Keeping it here for now as an experimental thing because all the terrain work is still very much in progress.